services/requester/requester.go

```diff
@@ -358,42 +362,113 @@ func (e *EVM) Call(
+	// given limit, but we probably don't want to return the lowest possible gas
+	// limit for these cases anyway.
+	failingGasLimit = result.GasConsumed - 1
+
+	// There's a fairly high chance for the transaction to execute successfully
+	// with gasLimit set to the first execution's GasConsumed + GasRefund.
+	// Explicitly check that gas amount and use as a limit for the binary search.
+	optimisticGasLimit := (result.GasConsumed + result.GasRefund + gethParams.CallStipend) * 64 / 63
+	if optimisticGasLimit < passingGasLimit {
+		tx.Gas = optimisticGasLimit
+		result, err = e.dryRunTx(tx, from, height, stateOverrides)
+		if err != nil {
+			// This should not happen under normal conditions since if we make it this far the
+			// transaction had run without error at least once before.
+			return 0, err
+		}
+		resultSummary := result.ResultSummary()
+		if resultSummary.ErrorCode == evmTypes.ExecutionErrCodeOutOfGas {
+			failingGasLimit = optimisticGasLimit
+		} else {
+			passingGasLimit = optimisticGasLimit
+		}
+	}
 
-	if result.Successful() {
-		// As mentioned in https://github.com/ethereum/EIPs/blob/master/EIPS/eip-150.md#specification
-		// Define "all but one 64th" of N as N - floor(N / 64).
-		// If a call asks for more gas than the maximum allowed amount
-		// (i.e. the total amount of gas remaining in the parent after subtracting
-		// the gas cost of the call and memory expansion), do not return an OOG error;
-		// instead, if a call asks for more gas than all but one 64th of the maximum
-		// allowed amount, call with all but one 64th of the maximum allowed amount of
-		// gas (this is equivalent to a version of EIP-901 plus EIP-1142).
-		// CREATE only provides all but one 64th of the parent gas to the child call.
-		result.GasConsumed = AddOne64th(result.GasConsumed)
-
-		// Adding `gethParams.SstoreSentryGasEIP2200` is needed for this condition:
-		// https://github.com/onflow/go-ethereum/blob/master/core/vm/operations_acl.go#L29-L32
-		result.GasConsumed += gethParams.SstoreSentryGasEIP2200
+	// Binary search for the smallest gas limit that allows the tx to execute successfully.
+	for failingGasLimit+1 < passingGasLimit {
+		// It is a bit pointless to return a perfect estimation, as changing
+		// network conditions require the caller to bump it up anyway. Since
+		// wallets tend to use 20-25% bump, allowing a small approximation
+		// error is fine (as long as it's upwards).
+		if float64(passingGasLimit-failingGasLimit)/float64(passingGasLimit) < estimateGasErrorRatio {
+			break
+		}
+		mid := (passingGasLimit + failingGasLimit) / 2
+		if mid > failingGasLimit*2 {
+			// Most txs don't need much higher gas limit than their gas used, and most txs don't
+			// require near the full block limit of gas, so the selection of where to bisect the
+			// range here is skewed to favor the low side.
+			mid = failingGasLimit * 2
+		}
+		tx.Gas = mid
+		result, err = e.dryRunTx(tx, from, height, stateOverrides)
```

Comment:
It would be interesting to see a metric of how many iterations we make on average to get the `passingGasLimit` maybe something we can add in the  future.

Reply:
That's a good idea :+1: